[Feature]  Add ENet

ETTR123 opened the pull request.

ETTR123 committed:
README.md

ETTR123 committed:
git add README.md

ETTR123 committed:
README.md

ETTR123 committed:
paddleseg/models/enet.py

ETTR123 committed:
paddle/models/enet/py

ETTR123 committed:
README

ETTR123 committed:
enet_cityscapes_1024x512_adam_0.002_80k.yml

ETTR123 committed:
__init__.py

ETTR123 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleSeg into new_branch

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/PaddleSeg?pullRequest=1573) <br/>All committers have signed the CLA.

shiyutang commented on .pre-commit-config.yaml:
```diff
@@ -8,7 +9,7 @@
 
 -   repo: https://github.com/pre-commit/pre-commit-hooks
-    sha: a11d9314b22d8f8c7556443875b731ef05965464
+    rev: a11d9314b22d8f8c7556443875b731ef05965464
```
why do you need to change this

ETTR123 replied:
Done.

pre-commit 2.15.0

在执行pre-commit命令时会出现如下警告：
[WARNING] Unexpected key(s) present on https://github.com/pre-commit/pre-commit-hooks: sha


[ERROR] Your pre-commit configuration is unstaged.
`git add .pre-commit-config.yaml` to fix this.
执行'git add .pre-commit-config.yaml'后会自动用‘rev’替换‘sha’

shiyutang commented on README.md:
````diff
@@ -197,3 +198,4 @@ If you find our project useful in your research, please consider citing:
     year={2019}
 }
 ```
+
````
please do not modify this file

ETTR123 replied:
Done.

shiyutang commented on configs/enet/README.md:
```diff
@@ -0,0 +1,14 @@
+
+| Model | Backbone | Resolution | Training Iters | mIoU | mIoU (flip) | mIoU (ms+flip) | Links |
+|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
+|ENet|-|1024x512|80000|-|-|-|[model](-) \| [log](-) \| [vdl](-)|
+|ENet|-|1024x512|120000|-|-|-|[model](-) \| [log](-) \| [vdl](-)|
```
plz add miou result on your model

ETTR123 replied:
Done

shiyutang commented on configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml:
```diff
@@ -0,0 +1,55 @@
+iters: 80000
```
if you only change the training iter, plz elaborate on why this is necessary. Normally, we do as the original implementation

ETTR123 replied:
Done.

shiyutang replied:
plz choose the result of 120k rather than 80k, it has better result and training iters is closer to the original implementation

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+                            output_size[2:],
+                            mode='bilinear',
+                            align_corners=self.align_corners)
+        # # else:
+        # #     if len(output_size) == len(self.kernel_size) + 2:
+        # #         output_size = output_size[2:]
+        # t = str(input.dtype).lower().strip().split(".")[-1]
+        # # print(t)
+        # t = TYPE_MAPPER[t]
+        # out = paddle.zeros(output_size, dtype=t)
+        # flatten_out = paddle.flatten(out)
+        # pdb.set_trace()
+        # for i in range(indices.shape[0]):
+        #     for j in range(indices.shape[1]):
+        #         for k in range(indices.shape[2]):
+        #             for m in range(indices.shape[3]):
+        #                 indices[i, j, k, m] = (out.shape[1] * out.shape[2] * out.shape[3]) * i + \
+        #                                       (out.shape[2] * out.shape[3]) * j + indices[i, j, k, m]
+        # flatten_indices = paddle.flatten(indices)
+        # flatten_input = paddle.flatten(input)
+        # for i in range(flatten_indices.shape[0]):
+        #     flatten_out[flatten_indices[i].tolist()] = flatten_input[i].tolist()
+        # out = paddle.reshape(flatten_out, out.shape)
```
plz remove these invalid code

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+
+
+class InitialBlock(nn.Layer):
+    """The initial block is composed of two branches:
+    1. a main branch which performs a regular convolution with stride 2;
+    2. an extension branch which performs max-pooling.
+    Doing both operations in parallel and concatenating their results
+    allows for efficient downsampling and expansion. The main branch
+    outputs 13 feature maps while the extension branch outputs 3, for a
+    total of 16 feature maps after concatenation.
+    Keyword arguments:
+    - in_channels (int): the number of input channels.
+    - out_channels (int): the number output channels.
+    - kernel_size (int, optional): the kernel size of the filters used in
+    the convolution layer. Default: 3.
+    - padding (int, optional): zero-padding added to both sides of the
+    input. Default: 0.
+    - bias (bool, optional): Adds a learnable bias to the output if
+    ``True``. Default: False.
+    - relu (bool, optional): When ``True`` ReLU is used as the activation
+    function; otherwise, PReLU is used. Default: True.
```
the format is not correct, plz refer to other model implementations

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+        # Extension branch
+        self.ext_branch = nn.MaxPool2D(3, stride=2, padding=1)
+
+        # Initialize batch normalization to be used after concatenation
```
Annotations like this is not necessary, plz remove it

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+
+
+class RegularBottleneck(nn.Layer):
+    """Regular bottlenecks are the main building block of ENet.
+    Main branch:
+    1. Shortcut connection.
+    Extension branch:
+    1. 1x1 convolution which decreases the number of channels by
+    ``internal_ratio``, also called a projection;
+    2. regular, dilated or asymmetric convolution;
+    3. 1x1 convolution which increases the number of channels back to
+    ``channels``, also called an expansion;
+    4. dropout as a regularizer.
+    Keyword arguments:
+    - channels (int): the number of input and output channels.
+    - internal_ratio (int, optional): a scale factor applied to
+    ``channels`` used to compute the number of
+    channels after the projection. eg. given ``channels`` equal to 128 and
+    internal_ratio equal to 2 the number of channels after the projection
+    is 64. Default: 4.
+    - kernel_size (int, optional): the kernel size of the filters used in
+    the convolution layer described above in item 2 of the extension
+    branch. Default: 3.
+    - padding (int, optional): zero-padding added to both sides of the
+    input. Default: 0.
+    - dilation (int, optional): spacing between kernel elements for the
+    convolution described in item 2 of the extension branch. Default: 1.
+    asymmetric (bool, optional): flags if the convolution described in
+    item 2 of the extension branch is asymmetric or not. Default: False.
+    - dropout_prob (float, optional): probability of an element to be
+    zeroed. Default: 0 (no dropout).
+    - bias (bool, optional): Adds a learnable bias to the output if
+    ``True``. Default: False.
+    - relu (bool, optional): When ``True`` ReLU is used as the activation
+    function; otherwise, PReLU is used. Default: True.
+    """
```
format problem

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+
+
+class UpsamplingBottleneck(nn.Layer):
+    """The upsampling bottlenecks upsample the feature map resolution using max
+    pooling indices stored from the corresponding downsampling bottleneck.
+    Main branch:
+    1. 1x1 convolution with stride 1 that decreases the number of channels by
+    ``internal_ratio``, also called a projection;
+    2. max unpool layer using the max pool indices from the corresponding
+    downsampling max pool layer.
+    Extension branch:
+    1. 1x1 convolution with stride 1 that decreases the number of channels by
+    ``internal_ratio``, also called a projection;
+    2. transposed convolution (by default, 3x3);
+    3. 1x1 convolution which increases the number of channels to
+    ``out_channels``, also called an expansion;
+    4. dropout as a regularizer.
+    Keyword arguments:
+    - in_channels (int): the number of input channels.
+    - out_channels (int): the number of output channels.
+    - internal_ratio (int, optional): a scale factor applied to ``in_channels``
+     used to compute the number of channels after the projection. eg. given
+     ``in_channels`` equal to 128 and ``internal_ratio`` equal to 2 the number
+     of channels after the projection is 64. Default: 4.
+    - dropout_prob (float, optional): probability of an element to be zeroed.
+    Default: 0 (no dropout).
+    - bias (bool, optional): Adds a learnable bias to the output if ``True``.
+    Default: False.
+    - relu (bool, optional): When ``True`` ReLU is used as the activation
+    function; otherwise, PReLU is used. Default: True.
+    """
+    def __init__(self,
```
same here

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+
+
+class DownsamplingBottleneck(nn.Layer):
+    """Downsampling bottlenecks further downsample the feature map size.
+    Main branch:
+    1. max pooling with stride 2; indices are saved to be used for
+    unpooling later.
+    Extension branch:
+    1. 2x2 convolution with stride 2 that decreases the number of channels
+    by ``internal_ratio``, also called a projection;
+    2. regular convolution (by default, 3x3);
+    3. 1x1 convolution which increases the number of channels to
+    ``out_channels``, also called an expansion;
+    4. dropout as a regularizer.
+    Keyword arguments:
+    - in_channels (int): the number of input channels.
+    - out_channels (int): the number of output channels.
+    - internal_ratio (int, optional): a scale factor applied to ``channels``
+    used to compute the number of channels after the projection. eg. given
+    ``channels`` equal to 128 and internal_ratio equal to 2 the number of
+    channels after the projection is 64. Default: 4.
+    - return_indices (bool, optional):  if ``True``, will return the max
+    indices along with the outputs. Useful when unpooling later.
+    - dropout_prob (float, optional): probability of an element to be
+    zeroed. Default: 0 (no dropout).
+    - bias (bool, optional): Adds a learnable bias to the output if
+    ``True``. Default: False.
+    - relu (bool, optional): When ``True`` ReLU is used as the activation
+    function; otherwise, PReLU is used. Default: True.
+    """
```
same here

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+        # asymmetric convolution, followed by another 1x1 convolution, and,
+        # finally, a regularizer (spatial dropout). Number of channels is constant.
+
+        # 1x1 projection convolution
```
remove this plz

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+            utils.load_pretrained_model(self, self.pretrained)
+
+
+class MaxUnpool2D(nn.Layer):
+    def __init__(self,
+                 kernel_size,
+                 stride=None,
+                 padding=0,
+                 align_corners=False):
+        super(MaxUnpool2D, self).__init__()
+
+        if isinstance(stride, int):
+            self.kernel_size = (kernel_size, kernel_size)
+        else:
+            self.kernel_size = kernel_size
+        if stride is None:
+            self.stride = self.kernel_size
+        else:
+            if isinstance(stride, int):
+                self.stride = (stride, stride)
+            else:
```
paddle have implemented maxunpool2d, plz refer to that implementation

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,759 @@
+        self.init_weight()
+
+    def forward(self, x):
+        # Initial block
+        # import pdb;pdb.set_trace()
+        # pdb.set_trace()
```
remove unnecessary annotation plz.

ETTR123 replied:
Done.

shiyutang commented on configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml:
```diff
@@ -0,0 +1,55 @@
+loss:
+  types:
+    - type: CrossEntropyLoss
+      """
+      weight为提取出来的cityscapes类别权重，不需要使用即可得到原论文结果
+      weight: [3.36023803, 14.0141461 ,  4.98533063, 39.23698918,
+      36.4913192 , 32.89015894, 46.28576252, 40.68635137,  6.69966664,
+      33.53070196, 18.52586764, 32.96432581, 47.67463868, 12.68382531,
+      45.19880984, 45.78221583, 45.81911329, 48.40295208, 42.74750531]
+      """
```
remove this plz

ETTR123 replied:
Done.

ETTR123 committed:
.pre-commit-config.yaml

ETTR123 committed:
all

ETTR123 committed:
paddleseg/models/enet.py

ETTR123 committed:
README.md

ETTR123 committed:
Update .pre-commit-config.yaml

ETTR123 committed:
Update enet_cityscapes_1024x512_adam_0.002_80k.yml

ETTR123 committed:
Update enet.py

MaxUnpool2D和Paddle.nn.MaxUnPool2D有不同，主要体现在传入forward的函数参数上，因此得使用自定义的MaxUnpool2D。

ETTR123 committed:
Update README.md

ETTR123 committed:
Update enet.py

shiyutang commented on configs/enet/README.md:
```diff
@@ -0,0 +1,13 @@
+
+| Model | Backbone | Resolution | Training Iters | mIoU | mIoU (flip) | mIoU (ms+flip) | Links |
+|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
+|ENet|-|1024x512|80000|-|-|58.3%|[model](-) \| [log](-) \| [vdl](-)|
```
plz write the result of 120k and make sure of its test configuration, that is whether it uses multi-scale & filp to get the result.

shiyutang commented on configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml:
```diff
@@ -0,0 +1,48 @@
+    - type: Normalize
+  mode: val
+
+
```
remove redundant space line plz.

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+    "fp64": "float64",
+    "float32": "float32",
+    "float64": "float64"
+}
```
remove this plz, you did not use it elsewhere

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pdb
```
Remove pdb plz. it is a debugger.

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+
+        # Main branch - As stated above the number of output channels for this
+        # branch is the total minus 3, since the remaining channels come from
+        # the extension branch
```
remove this plz

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+        super(RegularBottleneck, self).__init__()
+
+        # Check in the internal_scale parameter is within the expected range
+        # [1, channels]
```
Annotations like this should remove. plz check line 376, line 386-line388 and etc.

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+    Downsampling bottlenecks further downsample the feature map size.
+    Main branch:
+    1. max pooling with stride 2; indices are saved to be used for
+    unpooling later.
```
Add necessary spaceline for reading plz

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+    2. regular convolution (by default, 3x3);
+    3. 1x1 convolution which increases the number of channels to
+    ``out_channels``, also called an expansion;
+    4. dropout as a regularizer.
```
Add necessary spaceline for reading plz.

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+        # Before concatenating, check if main is on the CPU or GPU and
+        # convert padding accordingly
+        # if main.is_cuda:
+        # padding = padding.cuda()
```
remove annotated code plz.

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+                mode='bilinear',
+                align_corners=self.align_corners)
+
+        return out
```
maxunpool is implemented in paddle2.2.0 plz replace this class with paddle implementation.https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/MaxUnPool2D_cn.html

ETTR123 replied:
Done.

shiyutang commented on paddleseg/models/enet.py:
```diff
@@ -0,0 +1,703 @@
+class UpsamplingBottleneck(nn.Layer):
+    """
+    The upsampling bottlenecks upsample the feature map resolution using max
+    pooling indices stored from the corresponding downsampling bottleneck.
```
Add spaceline plz

ETTR123 replied:
Done.

ETTR123 committed:
Merge pull request #1 from PaddlePaddle/develop

Develop

ETTR123 committed:
Merge pull request #2 from ETTR123/new_branch

New branch

ETTR123 committed:
Update enet_cityscapes_1024x512_adam_0.002_80k.yml

ETTR123 committed:
update

ETTR123 committed:
Update enet.py

ETTR123 committed:
Update enet.py

ETTR123 committed:
Merge branch 'PaddlePaddle:develop' into develop

ETTR123 committed:
Merge pull request #3 from PaddlePaddle/develop

Develop

ETTR123 committed:
Delete README.md

ETTR123 committed:
Merge branch 'new_branch' of https://github.com.cnpmjs.org/ETTR123/PaddleSeg into new_branch

ETTR123 committed:
add test_tipc/configs/encnet_small

ETTR123 committed:
Merge branch 'PaddlePaddle:develop' into develop

ETTR123 committed:
Merge pull request #4 from ETTR123/new_branch

add test_tipc/configs/encnet_small

ETTR123 committed:
add encnet

ETTR123 committed:
Merge branch 'PaddlePaddle:develop' into develop

ETTR123 committed:
Merge pull request #5 from ETTR123/new_branch

add encnet

ETTR123 committed:
Update train_infer_python.txt

ETTR123 committed:
Create output

ETTR123 committed:
Delete output

ETTR123 committed:
Create results_python.log

ETTR123 committed:
Delete results_python.log

ETTR123 committed:
Create results_python.log

ETTR123 committed:
Merge pull request #6 from ETTR123/new_branch

New branch

ETTR123 committed:
Update train_infer_python.txt

ETTR123 committed:
Merge pull request #7 from ETTR123/develop

update

ETTR123 committed:
add fastfcn

ETTR123 committed:
update

ETTR123 committed:
update test_tipc/output/fastfcn_small/results_python.log

ETTR123 committed:
add dmnet

ETTR123 committed:
Create results_python.log
